File: color.ts
type Rgb = { r: number; g: number; b: number };

function rgbOf(r: number, g: number, b: number): Rgb {
  return { r, g, b };
}

function darken({ r, g, b }: Rgb): Rgb {
  let scale = 0.5;
  return rgbOf(scale * r, scale * g, scale * b);
  // TS 4.1 should allow for types like `Tuple<number, 3>`.
  // type Num3 = [number, number, number];
  // return rgbOf(...([r, g, b].map((x) => scale * x) as Num3));
}

// Pretend I deep froze this or something.
const namedColors: Record<string, Rgb> = {
  red: rgbOf(1, 0, 0),
  yellow: rgbOf(1, 1, 0),
  blue: rgbOf(0, 0, 1),
};

function main() {
  let names = ["red", "yellow", "blue"];
  let rgbs = names.map((name) => namedColors[name]);
  let darks = rgbs.map(darken);
  for (let i = 0; i < names.length; i += 1) {
    console.log(names[i], rgbs[i], darks[i]);
  }
}

main();

File: color.scala
case class Rgb(r: Double, g: Double, b: Double)

def darken(rgb: Rgb): Rgb = {
  val scale = 0.5
  Rgb(scale * rgb.r, scale * rgb.g, scale * rgb.b)
  // rgb match {
  //   case Rgb(r, g, b) => Rgb(scale * r, scale * g, scale * b)
  // }
}

val NamedColors = Map(
  "red" -> Rgb(1, 0, 0),
  "yellow" -> Rgb(1, 1, 0),
  "blue" -> Rgb(0, 0, 1),
)

@main def color = {
  val names = Vector[String]("red", "yellow", "blue")
  val rgbs = names.map(name => NamedColors(name))
  val darks = rgbs.map(darken)
  for (i <- 0 until names.length) {
    println(s"${names(i)} ${rgbs(i)} ${darks(i)}")
  }
}
